pkg/credentials/credentials.go

```diff
@@ -283,13 +287,45 @@ func defaultValidationFunc(creds map[string]string) error {
 }
 
 func openstackValidationFunc(creds map[string]string) error {
-	for k, v := range creds {
-		if k == OpenStackTenantID || k == OpenStackTenantName {
-			continue
+	alwaysRequired := []string{OpenStackAuthURL, OpenStackDomainName, OpenStackRegionName}
+
+	for _, key := range alwaysRequired {
+		if v, ok := creds[key]; !ok || len(v) == 0 {
+			return errors.Errorf("key %v is required but is not present", key)
 		}
-		if len(v) == 0 {
-			return errors.Errorf("key %v is required but isn't present", k)
+	}
+
+	var (
+		appCredsInUse  bool
+		userCredsInUse bool
+	)
+
+	if v, ok := creds[OpenStackApplicationCredentialID]; ok && len(v) != 0 {
+		if v, ok := creds[OpenStackApplicationCredentialSecret]; !ok || len(v) == 0 {
+			return errors.Errorf("key %v is required if %v is present", OpenStackApplicationCredentialSecret, OpenStackApplicationCredentialID)
 		}
+		appCredsInUse = true
+	}
```

Comment:
What if we have the other way -- Credential Secret provided, but not the Credential ID? We'll pass the validation and deploy the Credential Secret, but it wouldn't be useful without the ID (unless I'm mistaken). I think we should catch this case as well.

Reply:
Should be addressed now.